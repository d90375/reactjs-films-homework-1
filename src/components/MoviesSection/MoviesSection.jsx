import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MovieList from './components/MovieList';
import ModalWindow from './components/ModalWindow/ModalWindow';
import Spinner from '../Spinner';
import FilterTabs from './components/FilterTabs/FilterTabs';

import styles from './MoviesSection.scss';

class MoviesSection extends Component {
  async componentDidMount() {
    const { fetchGenres } = this.props;
    await fetchGenres();
    const { condition, genres, fetchMovies } = this.props;
    fetchMovies(condition, genres);
  }

  fetchByFilter = async (filter) => {
    const { setMoviesCondition } = this.props;
    await setMoviesCondition(filter);
    const { condition, genres, fetchMovies } = this.props;
    fetchMovies(condition, genres);
  };

  render() {
    const {
      error, pending, isModalWindow, trailer, removeTrailerInfo,
      trailerPending, trailerError, genres,
    } = this.props;

    if (error) {
      return (
        <div className={styles.container}>{`Error! ${error.message}`}</div>
      );
    }

    if (pending) {
      return (
        <div className={styles.container}>
          <Spinner />
        </div>
      );
    }

    return (
      <section className={styles.container}>
        {isModalWindow ? (
          <ModalWindow
            trailer={trailer}
            removeTrailerInfo={removeTrailerInfo}
            trailerPending={trailerPending}
            trailerError={trailerError}
          />
        )
          : null }
        <FilterTabs genres={genres} fetchByFilter={this.fetchByFilter} />
        <MovieList />
      </section>
    );
  }
}

MoviesSection.propTypes = {
  error: PropTypes.shape({
    message: PropTypes.string.isRequired,
  }),
  pending: PropTypes.bool.isRequired,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  condition: PropTypes.string.isRequired,
  fetchGenres: PropTypes.func.isRequired,
  fetchMovies: PropTypes.func.isRequired,
  removeTrailerInfo: PropTypes.func.isRequired,
  trailer: PropTypes.shape({}),
  trailerError: PropTypes.shape({}),
  trailerPending: PropTypes.bool.isRequired,
  isModalWindow: PropTypes.bool.isRequired,
  setMoviesCondition: PropTypes.func.isRequired,
};

MoviesSection.defaultProps = {
  error: null,
  trailer: null,
  trailerError: null,
};

export default MoviesSection;
