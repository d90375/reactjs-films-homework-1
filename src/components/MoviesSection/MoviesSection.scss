.container {
  padding: 70px 70px 0 70px;
  background-color: rgb(231, 231, 231);
}
