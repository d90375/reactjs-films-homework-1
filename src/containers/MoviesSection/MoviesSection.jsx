import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import MovieList from '../../components/MovieList';
import Modal from '../../components/Modal';
import Spinner from '../../components/Spinner';
import FilterTabs from '../../components/FilterTabs';
import styles from './MoviesSection.scss';

class MoviesSection extends Component {
  async componentDidMount() {
    const { location: { search } } = this.props;
    const params = new URLSearchParams(search);
    const filter = params.get('filter');
    const genreId = params.get('genreId');
    const query = params.get('search');
    const {
      fetchMovies, setMoviesCondition, setSearchQuery, deleteSearchQuery,
    } = this.props;

    if (filter || genreId) {
      deleteSearchQuery();
      await setMoviesCondition(filter || genreId);
    }

    if (query) {
      setSearchQuery(query);
      await setMoviesCondition('Search');
    }

    const { condition } = this.props;
    fetchMovies(condition, query);
  }

  componentDidUpdate(prevProps) {
    const { location: { search } } = this.props;
    if (search !== prevProps.location.search) {
      const params = new URLSearchParams(search);
      const filter = params.get('filter');
      const genreId = params.get('genreId');

      if (filter) {
        this.fetchByFilter(filter);
      }

      if (genreId) {
        this.fetchByFilter(genreId);
      }
    }
  }

  fetchByFilter = async (filter) => {
    const { setMoviesCondition, deleteSearchQuery } = this.props;
    deleteSearchQuery();
    await setMoviesCondition(filter);
    const { condition, genres, fetchMovies } = this.props;
    fetchMovies(condition, genres);
  };

  render() {
    const {
      error, isLoading, isModalOpened, trailer, removeTrailerInfo,
      trailerIsLoading, trailerError, genres, condition, movies,
      fetchTrailer, removeDetailsInfo, setMoviesCondition, history,
    } = this.props;

    if (error) {
      return (
        <div className={styles.container}>{`Error! ${error.message}`}</div>
      );
    }

    if (isLoading) {
      return (
        <div className={styles.container}>
          <FilterTabs
            name="Tabs"
            genres={genres}
            fetchByFilter={this.fetchByFilter}
            condition={condition}
          />
          <Spinner />
        </div>
      );
    }

    return (
      <section className={styles.container}>
        {isModalOpened ? (
          <Modal
            trailer={trailer}
            removeTrailerInfo={removeTrailerInfo}
            trailerIsLoading={trailerIsLoading}
            trailerError={trailerError}
          />
        )
          : null }
        <FilterTabs
          name="Tabs"
          genres={genres}
          fetchByFilter={this.fetchByFilter}
          condition={condition}
          historyPush={history.push}
        />
        <MovieList
          movies={movies}
          fetchTrailer={fetchTrailer}
          removeDetailsInfo={removeDetailsInfo}
          setMoviesCondition={setMoviesCondition}
        />
      </section>
    );
  }
}

MoviesSection.propTypes = {
  error: PropTypes.shape({
    message: PropTypes.string.isRequired,
  }),
  isLoading: PropTypes.bool.isRequired,
  movies: PropTypes.arrayOf(PropTypes.object).isRequired,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  condition: PropTypes.string.isRequired,
  fetchMovies: PropTypes.func.isRequired,
  removeTrailerInfo: PropTypes.func.isRequired,
  trailer: PropTypes.shape({}),
  trailerError: PropTypes.shape({}),
  trailerIsLoading: PropTypes.bool.isRequired,
  fetchTrailer: PropTypes.func.isRequired,
  isModalOpened: PropTypes.bool.isRequired,
  setMoviesCondition: PropTypes.func.isRequired,
  removeDetailsInfo: PropTypes.func.isRequired,
  setSearchQuery: PropTypes.func.isRequired,
  deleteSearchQuery: PropTypes.func.isRequired,
  location: PropTypes.shape({
    search: PropTypes.string,
  }),
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
};

MoviesSection.defaultProps = {
  error: null,
  trailer: null,
  trailerError: null,
  location: PropTypes.shape({
    search: null,
  }),
};

export default withRouter(MoviesSection);
