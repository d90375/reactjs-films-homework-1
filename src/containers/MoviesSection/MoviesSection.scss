.container {
  min-height: calc(100vh - 210px);
  padding: 70px 70px 0 70px;
  background-color: rgb(231, 231, 231);
}
