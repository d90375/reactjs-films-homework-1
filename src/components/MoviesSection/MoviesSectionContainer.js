import { connect } from 'react-redux';
import { getMoviesPending, getMoviesError, getMoviesCondition } from '../../modules/movies/moviesSelector';
import { getGenres } from '../../modules/genres/genresSelector';
import {
  getTrailer, getTrailerPending, getTrailerError, getModalWindow,
} from '../../modules/trailer/trailerSelector';
import { fetchMovies, setMoviesCondition } from '../../modules/movies/moviesAction';
import { fetchGenres } from '../../modules/genres/genresAction';
import { removeTrailerInfo } from '../../modules/trailer/trailerAction';
import MoviesSection from './MoviesSection';

const mapStateToProps = (state) => ({
  error: getMoviesError(state),
  pending: getMoviesPending(state),
  condition: getMoviesCondition(state),
  genres: getGenres(state),
  isModalWindow: getModalWindow(state),
  trailer: getTrailer(state),
  trailerPending: getTrailerPending(state),
  trailerError: getTrailerError(state),
});


const mapDispatchToProps = {
  fetchMovies,
  fetchGenres,
  removeTrailerInfo,
  setMoviesCondition,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(MoviesSection);
