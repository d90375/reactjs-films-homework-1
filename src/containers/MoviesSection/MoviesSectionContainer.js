import { connect } from 'react-redux';
import {
  getMoviesIsLoading, getMoviesError, getMoviesCondition, getMovies, getGenres,
} from '../../modules/movies/moviesSelector';
import {
  getTrailer, getTrailerIsLoading, getTrailerError, getModalOpened,
} from '../../modules/trailer/trailerSelector';
import { fetchTrailer, removeTrailerInfo } from '../../modules/trailer/trailerAction';
import {
  fetchMovies, setMoviesCondition, setSearchQuery, deleteSearchQuery,
} from '../../modules/movies/moviesAction';
import { removeDetailsInfo } from '../../modules/details/detailsAction';

import MoviesSection from './MoviesSection';

export const mapStateToProps = (state) => ({
  error: getMoviesError(state),
  isLoading: getMoviesIsLoading(state),
  movies: getMovies(state),
  condition: getMoviesCondition(state),
  genres: getGenres(state),
  isModalOpened: getModalOpened(state),
  trailer: getTrailer(state),
  trailerIsLoading: getTrailerIsLoading(state),
  trailerError: getTrailerError(state),
});


const mapDispatchToProps = {
  fetchMovies,
  fetchTrailer,
  removeTrailerInfo,
  setMoviesCondition,
  removeDetailsInfo,
  setSearchQuery,
  deleteSearchQuery,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(MoviesSection);
