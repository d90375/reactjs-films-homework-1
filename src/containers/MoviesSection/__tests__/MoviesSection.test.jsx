import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router, MemoryRouter, Route } from 'react-router-dom';
import { create } from 'react-test-renderer';
import MoviesSection from '../MoviesSection';

describe('MoviesSection tests', () => {
  describe('MoviesSection render', () => {
    const mockCallBack = jest.fn();

    const mockProps = {
      trailer: null,
      removeTrailerInfo: mockCallBack,
      trailerIsLoading: false,
      trailerError: null,
      condition: 'Trending',
      fetchTrailer: mockCallBack,
      fetchMovies: mockCallBack,
      setMoviesCondition: mockCallBack,
      removeDetailsInfo: mockCallBack,
      setSearchQuery: mockCallBack,
      deleteSearchQuery: mockCallBack,
    };

    it('MoviesSection renders correctly with closed modal window', () => {
      const genres = ['crime', 'actions'];
      const movies = [{ id: 123, title: 'title' }];

      const tree = create(
        <Router>
          <MoviesSection
            error={null}
            isLoading={false}
            isModalOpened={false}
            genres={genres}
            movies={movies}
            {...mockProps}
          />
        </Router>,
      );
      expect(tree).toMatchSnapshot();
    });

    it('MoviesSection renders correctly when modal window open', () => {
      const genres = ['crime', 'actions'];
      const movies = [{ id: 123, title: 'title' }];

      const tree = create(
        <Router>
          <MoviesSection
            error={null}
            isLoading={false}
            isModalOpened
            genres={genres}
            movies={movies}
            {...mockProps}
          />
        </Router>,
      );
      expect(tree).toMatchSnapshot();
    });

    it('MoviesSection renders correctly when movies is loading', () => {
      const genres = [];
      const movies = [];

      const tree = create(
        <Router>
          <MoviesSection
            error={null}
            isLoading
            isModalOpened={false}
            genres={genres}
            movies={movies}
            {...mockProps}
          />
        </Router>,
      );
      expect(tree).toMatchSnapshot();
    });

    it('MoviesSection renders correctly with error message', () => {
      const genres = [];
      const movies = [];
      const error = { message: 'error' };

      const tree = create(
        <Router>
          <MoviesSection
            error={error}
            isLoading={false}
            isModalOpened={false}
            genres={genres}
            movies={movies}
            {...mockProps}
          />
        </Router>,
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('ComponentDidMount', () => {
    const mockCallBack = jest.fn();

    const mockProps = {
      error: null,
      isLoading: false,
      isModalOpened: false,
      trailer: null,
      removeTrailerInfo: mockCallBack,
      trailerIsLoading: false,
      trailerError: null,
      genres: ['crime', 'actions'],
      condition: 'Trending',
      movies: [{ id: 123, title: 'title' }, { id: 110, title: 'title' }],
      fetchTrailer: mockCallBack,
      setSearchQuery: mockCallBack,
      deleteSearchQuery: mockCallBack,
      removeDetailsInfo: mockCallBack,
    };

    it('fetchMovies, fetchByFilter functions called', () => {
      const mockFetchMovies = jest.fn();
      const mockSetMoviesCondition = jest.fn();
      const tree = create(
        <Router>
          <MoviesSection
            fetchMovies={mockFetchMovies}
            setMoviesCondition={mockSetMoviesCondition}
            {...mockProps}
          />
        </Router>,
      );

      const filterTabs = tree.root.findByProps({ name: 'Tabs' });
      filterTabs.props.fetchByFilter('Trending');

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(1);
      expect(mockFetchMovies.mock.calls.length).toEqual(1);
    });

    it('setMoviesCondition functions called with filter', () => {
      const mockFetchMovies = jest.fn();
      const mockSetMoviesCondition = jest.fn();
      create(
        <MemoryRouter initialEntries={['/?filter=Trending']}>
          <Route path="/">
            <MoviesSection
              fetchMovies={mockFetchMovies}
              setMoviesCondition={mockSetMoviesCondition}
              {...mockProps}
            />
          </Route>
        </MemoryRouter>,
      );

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(1);
      expect(mockSetMoviesCondition).toHaveBeenCalledWith('Trending');
    });

    it('setMoviesCondition functions called with genreId', () => {
      const mockFetchMovies = jest.fn();
      const mockSetMoviesCondition = jest.fn();
      create(
        <MemoryRouter initialEntries={['/?genreId=16']}>
          <Route path="/">
            <MoviesSection
              fetchMovies={mockFetchMovies}
              setMoviesCondition={mockSetMoviesCondition}
              {...mockProps}
            />
          </Route>
        </MemoryRouter>,
      );

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(1);
      expect(mockSetMoviesCondition).toHaveBeenCalledWith('16');
    });

    it('setMoviesCondition functions called with "Search"', () => {
      const mockFetchMovies = jest.fn();
      const mockSetMoviesCondition = jest.fn();
      create(
        <MemoryRouter initialEntries={['/?search=sonic']}>
          <Route path="/">
            <MoviesSection
              fetchMovies={mockFetchMovies}
              setMoviesCondition={mockSetMoviesCondition}
              {...mockProps}
            />
          </Route>
        </MemoryRouter>,
      );

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(1);
      expect(mockSetMoviesCondition).toHaveBeenCalledWith('Search');
    });
  });

  /* describe('ComponentDidUpdate', () => {
    it('fetchByFilter functions called with filter', () => {
      const mockCallBack = jest.fn();
      const mockFetchMovies = jest.fn();
      const mockSetMoviesCondition = jest.fn();
      const mockSetSearchQuery = jest.fn();
      const mockDeleteSearchQuery = jest.fn();
      const genres = ['crime', 'actions'];
      const movies = [{ id: 123, title: 'title' }, { id: 110, title: 'title' }];
      const location = { path: ['/?filter=Coming%20soon', '/?filter=Trending'] };

      const node = document.createElement('div');
      const instance = ReactDOM.render(
        <MemoryRouter initialEntries={location.path} initialIndex={1}>
          <MoviesSection
            error={null}
            isLoading={false}
            isModalOpened={false}
            trailer={null}
            removeTrailerInfo={mockCallBack}
            trailerIsLoading={false}
            trailerError={null}
            genres={genres}
            condition="Coming Soon"
            movies={movies}
            fetchTrailer={mockCallBack}
            fetchMovies={mockFetchMovies}
            setMoviesCondition={mockSetMoviesCondition}
            setSearchQuery={mockSetSearchQuery}
            deleteSearchQuery={mockDeleteSearchQuery}
          />
        </MemoryRouter>, node,
      );

      // spyOn(MoviesSection, 'fetchByFilter');

      ReactDOM.render(
        <MemoryRouter initialEntries={location.path} initialIndex={0}>
          <MoviesSection
            error={null}
            isLoading={false}
            isModalOpened={false}
            trailer={null}
            removeTrailerInfo={mockCallBack}
            trailerIsLoading={false}
            trailerError={null}
            genres={genres}
            condition="Trending"
            movies={movies}
            fetchTrailer={mockCallBack}
            fetchMovies={mockFetchMovies}
            setMoviesCondition={mockSetMoviesCondition}
            setSearchQuery={mockSetSearchQuery}
            deleteSearchQuery={mockDeleteSearchQuery}
          />
        </MemoryRouter>, node,
      );
      expect(instance.fetchByFilter).toHaveBeenCalled();
    });
  }); */
});
