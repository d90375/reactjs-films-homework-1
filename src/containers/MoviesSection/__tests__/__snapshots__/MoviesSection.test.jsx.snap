// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesSection tests MoviesSection render MoviesSection renders correctly when modal window open 1`] = `
<section
  className="container"
>
  <Modal
    removeTrailerInfo={[MockFunction]}
    trailer={null}
    trailerError={null}
    trailerIsLoading={false}
  />
  <FilterTabs
    condition="Tranding"
    fetchByFilter={[Function]}
    genres={
      Array [
        "crime",
        "actions",
      ]
    }
    name="Tabs"
  />
  <MovieList
    fetchTrailer={[MockFunction]}
    movies={
      Array [
        Object {
          "id": 123,
        },
        Object {
          "id": 110,
        },
      ]
    }
  />
</section>
`;

exports[`MoviesSection tests MoviesSection render MoviesSection renders correctly when movies is loading 1`] = `
<div
  className="container"
>
  <Spinner />
</div>
`;

exports[`MoviesSection tests MoviesSection render MoviesSection renders correctly with closed modal window 1`] = `
<section
  className="container"
>
  <FilterTabs
    condition="Tranding"
    fetchByFilter={[Function]}
    genres={
      Array [
        "crime",
        "actions",
      ]
    }
    name="Tabs"
  />
  <MovieList
    fetchTrailer={[MockFunction]}
    movies={
      Array [
        Object {
          "id": 123,
        },
        Object {
          "id": 110,
        },
      ]
    }
  />
</section>
`;

exports[`MoviesSection tests MoviesSection render MoviesSection renders correctly with error message 1`] = `
<div
  className="container"
>
  Error! error
</div>
`;
